### PR TITLE
Added import action

### PR DESCRIPTION
**Pull request by dogversioning**
This adds the following:
- A new subcommand/action for importing previously exported studies
- Some alphasorting of large function blocks
- minor: made all custom exceptions children of CumulusLibraryException
- Added a custom argument param #247 

### Checklist
- [X] Consider if documentation (like in `docs/`) needs to be updated
- [X] Consider if tests should be added
- [X] Update template repo if there are changes to study configuration

### CODE REVIEW

File: cumulus_library/actions/cleaner.py [hunk 1]
```diff
@@ -76,7 +76,7 @@ def get_unprotected_stats_view_table(
 
 
 def clean_study(
-    manifest_parser: study_parser.StudyManifestParser,
+    manifest_parser: study_parser.StudyManifestParser | None,
```
Comment (dogversioning) [inline cumulus_library/actions/cleaner.py:79]: this allows the importer to not have to worry about setting up an empty parser.

File: cumulus_library/cli_parser.py
Comment (dogversioning) [inline cumulus_library/cli_parser.py:1]: changes in this file not related to import/custom args are just alphasorting.

File: cumulus_library/errors.py
Comment (dogversioning) [inline cumulus_library/errors.py:1]: These are also alphasorted in addition to now being child classes

File: cumulus_library/actions/importer.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import pandas
+
+from cumulus_library import base_utils, errors
+from cumulus_library.actions import cleaner
+from cumulus_library.template_sql import base_templates
+
+
+def create_table_from_parquet(archive, file, study_name, db, schema):
+    try:
+        parquet_path = pathlib.Path(archive.extract(file))
```
Comment (mikix) [inline cumulus_library/actions/importer.py:13]: Where does this extract to? `pwd`? Feels like we should point it at a temp folder/file.

File: cumulus_library/actions/importer.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        files = [file for file in files if file.endswith(".parquet")]
+    except zipfile.BadZipFile as e:
+        raise errors.StudyImportError(
+            f"File {archive_path} is not a valid archive."
+        ) from e
+    if len(files) == 0:
+        raise errors.StudyImportError(
+            f"File {archive_path} does not contain any tables."
+        )
+    study_name = files[0].split("__")[0]
```
Comment (mikix) [inline cumulus_library/actions/importer.py:56]: nit: there isn't a sanity check that there actually is a `__` in that filename. Not super likely, but the surrounding code is pretty paranoid.

File: tests/test_duckdb.py [hunk 4]
```diff
@@ -77,7 +77,6 @@ def test_duckdb_load_ndjson_dir(tmp_path):
         "A.Patient.ndjson": False,
         "1.Patient.ndjson": True,
         "Patient.ndjson": True,
-        "patient.ndjson": False,
```
Comment (mikix) [inline tests/test_duckdb.py:80]: RIP

File: tests/test_cli.py [hunk 5]
```diff
@@ -575,3 +575,29 @@ def test_cli_finds_study_from_manifest_prefix(tmp_path):
...
+
+@mock.patch.dict(os.environ, clear=True)
+@pytest.mark.parametrize(
+    "c_arg,raises",
+    [
+        ("foo:bar", does_not_raise()),
+        ("foo", pytest.raises(SystemExit)),
+    ],
+)
+def test_cli_custom_args(tmp_path, c_arg, raises):
```
Comment (mikix) [inline tests/test_cli.py:588]: nit: would be nice to confirm the value gets parsed and sent down all the way to the study config too.

**Review by mikix — APPROVED** [review]

File: cumulus_library/actions/cleaner.py (hunk 1 above)
Comment (dogversioning) [reply cumulus_library/actions/cleaner.py:79]: It does, but only if `prefix` is not supplied.

I missed updating the docstring here - i'll do that to clarify, and maybe and an error if both of these are null.

File: cumulus_library/cli_parser.py [hunk 6]
```diff
@@ -2,37 +2,34 @@
...
 
 
-def add_study_dir_argument(parser: argparse.ArgumentParser) -> None:
-    """Adds --study-dir arg to a subparser"""
+def add_custom_argument(parser: argparse.ArgumentParser) -> None:
     parser.add_argument(
-        "-s",
-        "--study-dir",
+        "-c",
+        "--custom-arg",
```
[the remark this answers is not in the document]
Comment (dogversioning) [reply cumulus_library/cli_parser.py:27]: i don't - i went back and forth on this several times and was never thrilled. i'll change this to option.

File: cumulus_library/databases.py [hunk 7]
```diff
@@ -168,6 +169,37 @@ def operational_errors(self) -> tuple[Exception]:
         """
         return ()
 
+    def col_parquet_types_from_pandas(self, field_types: list) -> list:
```
[the remark this answers is not in the document]
Comment (dogversioning) [reply cumulus_library/databases.py:172]: > Also... it's rude that Athena can't just grab the types from the parquet file it has sitting there.

GOD right?

I tried this with pyarrow and there was some aspect of it that i had an issue with, though i don't remember the exact details - something about the way I was accessing the types ended up in a very deep set of nested objects that felt like overkill, but perhaps i missed another option. i'll take a look at the fhir-support approach.

File: cumulus_library/actions/importer.py (hunk 2 above)
Comment (dogversioning) [reply cumulus_library/actions/importer.py:13]: good call, will change.

File: cumulus_library/actions/importer.py (hunk 3 above)
Comment (dogversioning) [reply cumulus_library/actions/importer.py:56]: what's a little more paranoia between friends? will add.

File: tests/test_cli.py (hunk 5 above)
Comment (dogversioning) [reply tests/test_cli.py:588]: fair point - i can mock something out

File: cumulus_library/actions/cleaner.py (hunk 1 above)
Comment (mikix) [reply cumulus_library/actions/cleaner.py:79]: Ah yeah doh - I missed its use :smile:

File: cumulus_library/databases.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mikix) [reply cumulus_library/databases.py:172]: Also, this is fine - like, we aren't wedding ourselves to this approach in a meaningful way - can always refactor later if we decide we don't like pandas.

[the remark this answers is not in the document]
Comment (dogversioning) [reply cumulus_library/databases.py:172]: Added a comment about this for posteritiy and leaving it as is.

**Pull request merged**